File: Dotnet/Dotnet.Mqtt/MqttService.cs
using MQTTnet;
using MQTTnet.Client;
using MQTTnet.Formatter;
using Microsoft.Extensions.Logging;

namespace Dotnet.Mqtt;

public enum MqttAction
{
    STATE,
    COMMAND,
    EVENT
}

public interface IMqttService
{
    event EventHandler<MqttApplicationMessageReceivedEventArgs>? OnMessageReceived;
    event EventHandler? OnConnected;
    event EventHandler? OnDisconnect;

    Task StartAsync(CancellationToken stoppingToken);
    Task StopAsync();

    Task Publish(MqttAction action, string target, string actorId, string messageType, byte[] bytes);

    void Subscribe(List<string> topics);
}

public class MqttService : IMqttService
{
    private readonly MqttSettings settings;
    private readonly ILogger<MqttService> logger;
    private IMqttClient? mqttClient;

    public MqttService(MqttSettings settings, ILogger<MqttService> logger)
    {
        this.settings = settings;
        this.logger = logger;

        if (settings == null)
        {
            throw new Exception("Missing appSettings.Mqtt");
        }

        if (string.IsNullOrEmpty(settings.Host))
        {
            logger.LogError("Missing appSettings.Mqtt.Host");
        }
    }

    public event EventHandler<MqttApplicationMessageReceivedEventArgs>? OnMessageReceived;
    public event EventHandler? OnConnected;
    public event EventHandler? OnDisconnect;

    [Obsolete] //TODO: fix this
    public async Task StartAsync(CancellationToken stoppingToken)
    {
        var factory = new MqttFactory();
        var optionsBuilder = new MqttClientOptionsBuilder()
            .WithTcpServer(settings.Host, settings.Port)
            .WithCredentials(settings.User, settings.Password)
            .WithClientId($"{settings.ClientId}-{DateTimeOffset.Now.ToUnixTimeSeconds()}")
            .WithKeepAlivePeriod(TimeSpan.FromSeconds(5))
            .WithProtocolVersion(MqttProtocolVersion.V500);

        if (settings.IsTlsEnabled == true)
        {
            MqttClientOptionsBuilderTlsParameters parameters = new()
            {
                AllowUntrustedCertificates = true,
                UseTls = true,
                IgnoreCertificateChainErrors = true,
                IgnoreCertificateRevocationErrors = true
            };

            optionsBuilder = optionsBuilder.WithTls(parameters);
        }

        var options = optionsBuilder.Build();


        mqttClient = factory.CreateMqttClient();
        logger.LogInformation("### ATTEMPTING MQTT CONNECTION ###");

        mqttClient.ConnectedAsync += e => 
        {
            logger.LogInformation("### CONNECTED WITH MQTT SERVER ###");

            OnConnected?.Invoke(this, new EventArgs());

            logger.LogInformation("### SUBSCRIBED ###");

            return Task.CompletedTask;
        };

        mqttClient.DisconnectedAsync += async e => 
        {
            logger.LogError("### DISCONNECTED FROM MQTT SERVER ###");

            await Task.Delay(TimeSpan.FromSeconds(settings.RetryIntervalInSec));

            try
            {
                await mqttClient.ConnectAsync(options, stoppingToken);
            }
            catch
            {
                logger.LogError("### RECONNECTING MQTT FAILED ###");
            }
        };

        mqttClient.ApplicationMessageReceivedAsync += e => {
            OnMessageReceived?.Invoke(this, e);
            return Task.CompletedTask;
        };

        await mqttClient.ConnectAsync(options, stoppingToken);
    }

    public async Task StopAsync()
    {
        OnDisconnect?.Invoke(this, new EventArgs());
        await mqttClient.DisconnectAsync();
    }

    public async Task Publish(MqttAction action, string target, string actorId, string messageType, byte[] bytes)
    {
        if (mqttClient != null && mqttClient.IsConnected)
        {
            try
            {
                var msg = new MqttApplicationMessageBuilder()
                    .WithTopic($"{action.ToString().ToLower()}/{target}/{actorId}/{messageType}")
                    .WithPayload(bytes)
                    .WithQualityOfServiceLevel(MQTTnet.Protocol.MqttQualityOfServiceLevel.AtLeastOnce)
                    .WithRetainFlag()
                    .Build();

                Console.WriteLine($"Sending:... {action}/{target}/{actorId}/{messageType}");

                await mqttClient.PublishAsync(msg);
            }
            catch (Exception e)
            {
                this.logger.LogError($"Mqtt Send {action.ToString().ToLower()}/{target}/{actorId}/{messageType} failed: {e.Message}");
            }
        }
    }

    public void Subscribe(List<string> topics)
    {
        if (mqttClient != null && mqttClient.IsConnected)
        {
            var subscriptionBuilder = new MqttClientSubscribeOptionsBuilder();

            foreach (var topic in topics)
            {
                subscriptionBuilder.WithTopicFilter(topic);
            }

            mqttClient.SubscribeAsync(subscriptionBuilder.Build());
        }
    }
}

File: Python/src/mqtt/mqtt_service.py
from enum import Enum
import logging
from typing import Any, Callable, List
import paho.mqtt.client as mqtt
from mqtt.mqtt_settings import MqttSettings

class MqttAction(Enum):
    STATE = "STATE"
    COMMAND = "COMMAND"
    EVENT = "EVENT"

class MqttService:
    def __init__(self, mqtt_settings: MqttSettings):
        self.logger = logging.getLogger(__name__)

        self.mqtt_settings = mqtt_settings

        self.on_connect_callback: Callable[[mqtt.Client, Any, Any, int], None]
        self.on_disconnect_callback: Callable[[mqtt.Client, Any, int], None]
        self.on_message_callback: Callable[[mqtt.MQTTMessage], None]

    def start_mqtt(self):
        print(f"Connecting to {self.mqtt_settings.host}:{self.mqtt_settings.port}")

        self.client = mqtt.Client(transport='websockets')

        self.client.username_pw_set(self.mqtt_settings.user, self.mqtt_settings.password)

        # Certificate is required when using TLS endpoint (mqtts, wss, ssl)!
        if self.mqtt_settings.is_tls_enabled:
            self.client.tls_set()

        self.client.on_connect = self.on_connect
        self.client.on_message = self.on_message
        self.client.on_disconnect = self.on_disconnect

        self.client.connect(self.mqtt_settings.host, port=self.mqtt_settings.port, keepalive=5,)

        self.client.loop_forever()

    def stop_mqtt(self):
        self.client.disconnect()

    def on_connect(self, client: mqtt.Client, userdata, flags, rc: int):
        if self.on_connect_callback is not None:
            self.on_connect_callback(client, rc)

    def on_message(self, client: mqtt.Client, userdata, msg: mqtt.MQTTMessage):
        if self.on_message_callback is not None:
            self.on_message_callback(msg)

    def on_disconnect(self, client: mqtt.Client, userdata, rc: int):
        #TODO: add reconnect logic here

        if self.on_disconnect_callback is not None:
            self.on_disconnect_callback()

    def publish(self, action: MqttAction, target: str, actor_id: str, message_namespace: str, message_type: str, payload: bytes):
        """Publishes a message to a specific MQTT topic."""
        if self.client.is_connected():
            try:
                topic = f"{action.value.lower()}/{target}/{actor_id}/{message_namespace}.{message_type}"

                # Publish the message
                result = self.client.publish(
                    topic=topic,
                    payload=payload,
                    qos=1, #qos=1 --> AtLeastOnce
                    retain=True
                )
                
                if result.rc != mqtt.MQTT_ERR_SUCCESS:
                    self.logger.error(f"Failed to publish message to {topic}")

            except Exception as e:
                self.logger.error(f"Publish to {topic} failed: {e}")
        else:
            self.logger.warning("MQTT client is not connected")

    def subscribe(self, topics: List[str]):
        """Subscribes to a list of topics."""
        if self.client.is_connected():
            for topic in topics:
                self.client.subscribe(topic)
        else:
            self.logger.warning("MQTT client is not connected")
